Recipe Rewrite (API Only)

SchnTgaiSpock opened the pull request:
## Description

The long awaited recipe rewrite (supersedes #4093)

## Proposed changes

In short, adds true shaped/shapeless crafting, tagged inputs, and json (de)serialization (custom recipes)

Full ADR [here](https://github.com/SchnTgaiSpock/Slimefun4/blob/api/recipe-rewrite-4/docs/adr/0002-recipe-rewrite.md)

## Related Issues (if applicable)
N/A

## Checklist
- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.20.*).
- [ ] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

SchnTgaiSpock committed:
add new api

SchnTgaiSpock commented:
I'm going to throw together a showcase addon sometime next week to facilitate testing since I don't want to migrate any Slimefun stuff in this PR.

SchnTgaiSpock committed:
remove printlns

SchnTgaiSpock committed:
more unit tests

SchnTgaiSpock committed:
fix test

SchnTgaiSpock committed:
more unit tests

SchnTgaiSpock committed:
update api

SchnTgaiSpock committed:
implement loading from resources, reloading/saving, some api changes

SchnTgaiSpock committed:
remove test recipe

SchnTgaiSpock committed:
make subdirectories easier

SchnTgaiSpock committed:
add backup

SchnTgaiSpock committed:
add save and restore recipe backups

SchnTgaiSpock committed:
Merge branch 'Slimefun:master' into api/recipe-rewrite-4

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+# 2. Recipe rewrite
```
initial review is just on the adr - thank you for this by the way, it makes large changes like this much better :)

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+  of recipes with this system.
+- Customizability: Server owners should be able to customize any and all
+  Slimefun recipes
+- Performance: Should not blow up any servers
```
```suggestion
- Performance: Should be on par or better than the current system.
```

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+the recipe should be matched to items in a multiblock/machine when crafting. The base ones are:
+
+- Shaped/Shapeless: Exactly the same as vanilla
+- Subset: How the current smeltery, etc. craft
```
can we expand and say what this actually is

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+`RecipeInput`s are a list of `RecipeInputItem`s plus a `MatchProcedure` -- how the inputs of
+the recipe should be matched to items in a multiblock/machine when crafting. The base ones are:
+
+- Shaped/Shapeless: Exactly the same as vanilla
```
```suggestion
- Shaped/Shapeless: Exactly the same as vanilla. Shaped recipes need to be crafted exactly as defined. Shapeless need to be in the same order but can move (e.g. 3 sticks down the y axis could be on any x axis)
```

(ngl shaped + shapeless are both wrong in mc...)

SchnTgaiSpock replied:
i think we have differing definitions of shaped/shapeless; for me:
- shaped means as long as the order and relative positions of the items are the same, they can be moved anywhere on the crafting grid and it will be crafted
- shapeless means as long as the all the specified items (and nothing extra) are there it will be crafted regardless of order

these are what mc does and i want to keep the naming consistent. i also don't see why they are 'wrong' per se, although ordered and orderless might have been a better name tbh

what you defined as shaped i think i would call smth like 'fixed', and as a player i dislike those kinds of recipes -- copper wire for example needs 3 copper ingots in the middle row, which means i can't shift click them in (to the top row), whereas if it were 'shaped' i could.

i probably wasn't clear enough that i wanted to move away from those recipes since they are pretty annoying as a player to use

ill add it in (albeit w/ different name), but i don't see a reason to use it over mc shaped

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+
+- Shaped/Shapeless: Exactly the same as vanilla
+- Subset: How the current smeltery, etc. craft
+- Shaped-flippable: The recipe can be flipped on the Y-axis
```
give an example

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+- Shaped/Shapeless: Exactly the same as vanilla
+- Subset: How the current smeltery, etc. craft
+- Shaped-flippable: The recipe can be flipped on the Y-axis
+- Shaped-rotatable: The recipe can be rotated (currently only 45deg, 3x3)
```
let's give a simple example

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+
+An `RecipeOutputItem`s controls how an output is generated when the recipe is
+crafted. It can be a single item (see `RecipeOutputItemStack`, `RecipeOutputSlimefunItem`),
+or a group of items each with a certain weight of being output (see `RecipeOutputGroup`).
```
can i return just multiple items in any way? 
also i'd be clear that this weighting is still outputing a single item, just with different chances - i didn't get what this was saying until the example below

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+
+### RecipeService
+
+This is the public interface for the recipe system, there are methods here to add,
```
i'd probably keep the whole registration naming/process rather than "add"

```suggestion
This is the public interface for the recipe system, there are methods here to register,
```

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+{
+    "input"?: RecipeInput
+    "output"?: RecipeOutput
+    "type": key | key[]
```
A recipe being multiple types is insteresting
```suggestion
    "type": NamespacedKey | NamespacedKey[]
```

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+    "input"?: RecipeInput
+    "output"?: RecipeOutput
+    "type": key | key[]
+    "energy"?: int
+    "crafting-time"?: int
```
I'm not fully sure we should have these at least top level
My thinking is around addons and custom properties, I'd rather not pollute top level with all these custom properties (and we should treat anything sf as a custom property too - we should essentially be the main addon)

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+    "type": key | key[]
+    "energy"?: int
+    "crafting-time"?: int
+    "permission-node"?: string | string[]
```
any reason to use - by the way here?
i feel like camelCase is much more common within json

SchnTgaiSpock replied:
no particular reason, ill change it to camelCase

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
````diff
@@ -0,0 +1,254 @@
+}
+```
+
+The recipe deserializer technically needs a `__filename` field, but it is
````
interesting 🤔

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
````diff
@@ -0,0 +1,254 @@
+`Recipe`
+
+```txt
+{
````
i don't see any like "id" field

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
````diff
@@ -0,0 +1,254 @@
+
+```txt
+{
+    "items": string | string[]
````
how do these string references work? if i want to point to a vanilla item or an sf item - what do i do there?
(may be worth giving examples for these)

SchnTgaiSpock replied:
added an example in comment below, ill also put it in the adr

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
````diff
@@ -0,0 +1,254 @@
+```txt
+{
+    "items": string | string[]
+    "key": {
+        [key: string]: RecipeInputItem
+    }
````
I don't get how this is meant to look today

SchnTgaiSpock replied:
each key in `key` is a character that you used in the `items` field
```
"input": {
    "items": [
        "123",
        " 4 ",
        "   "
    ],
    "key": {
        "1": "slimefun:nickel_ingot",
        "2": "slimefun:magnet",
        "3": "slimefun:cobalt_ingot",
        "4": "slimefun:battery"
    },
    "match": "slimefun:shaped"
},
```

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+    "key": {
+        [key: string]: RecipeInputItem
+    }
+    "match"?: key
```
```suggestion
    "match"?: NamespacedKey
```

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
````diff
@@ -0,0 +1,254 @@
+
+```txt
+{
+    "id": key
````
```suggestion
    "id": NamespacedKey
```

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+{
+    "id": key
+    "amount"?: int
+    "durability"?: int
```
why durability?

SchnTgaiSpock replied:
i remember some mods that had recipes like this, where you could put an item and some kind of tool in, and when you crafted, it would use up some dura on the tool (i *think* it was farmers delight but i could be misremembering)

i thought that would be cool to add, but if you think its out of scope or should just be a custom recipe ill remove it

SchnTgaiSpock replied:
yeah coming back 5 months later it seems unnecessary, removed it

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+    "amount"?: int
+    "durability"?: int
+} | {
+    "tag": key
```
```suggestion
    "tag": NamespacedKey
```

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+    "amount"?: int
+} | {
+    "group": RecipeInputItem[]
+    "weights"?: int[]
```
guessing weights have to line up index wise? If so, that's not great dx. I could add a new item and forget a weight - what happens then?

SchnTgaiSpock replied:
yeah good point, not sure why i did it like this

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+
+*In addition to those schemas, items can be in short form:
+
+- Single items: `<namespace>:<id>|<amount>`
+- Tags: `#<namespace>:<id>|<amount>`
```
tbh i'm not a big fan of this, i'd rather be more explicit. I can feel this getting unwieldy in the future

SchnTgaiSpock replied:
just to be clear, the short form is only used for items that are *just* an id (mc or sf) and amount

i think this would just be a lot cleaner when writing recipes, since most inputs won't have any other fields anyways

for example
```
{
    "id": "ELECTROMAGNET",
    "input": {
        "items": [
            "123",
            " 4 ",
            "   "
        ],
        "key": {
            "1": "slimefun:nickel_ingot",
            "2": "slimefun:magnet",
            "3": "slimefun:cobalt_ingot",
            "4": "slimefun:battery"
        },
        "match": "slimefun:shaped"
    },
    "output": {
        "items": [
            "slimefun:electromagnet"
        ]
    },
    "type": "slimefun:enhanced_crafting_table"
}
```
vs
```
{
    "id": "ELECTROMAGNET",
    "input": {
        "items": [
            "123",
            " 4 ",
            "   "
        ],
        "key": {
            "1": {
                "id": "slimefun:nickel_ingot"
            },
            "2": {
                "id": "slimefun:magnet"
            },
            "3": {
                "id": "slimefun:cobalt_ingot"
            },
            "4": {
                "id": "slimefun:battery"
            }
        },
        "match": "slimefun:shaped"
    },
    "output": {
        "items": [
            {
                "id": "slimefun:electromagnet"
            }
        ]
    },
    "type": "slimefun:enhanced_crafting_table"
}
```

i forgot to mention the `|<amount>` bit was optional if its just 1; ill edit that in.

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+the presence of a `class` field in the json, which should be the key of a
+custom deserializer registered with Slimefun's `RecipeService`.
+For custom serializers, override the `serialize` method on the subclass,
+and ensure they also add the `class` field
```
nts: come back to this class field

SchnTgaiSpock replied:
i might have another solution: registering a deserializer to work addon-wide, but ig there is still the issue of if addon A wants to depend on addon B and use B's deserializers

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+moved to `plugins/Slimefun/recipes/` (unless a file with its name already exists).
+
+Addons should do the same. (We recommend saving to
+`plugins/Slimefun/recipes/<your-addon-name>/` but it's not required).
```
wonder if we just make this required - otherwise i think we just have people overriding eachother.

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+
+### Stage 2
+
+On the first server tick, all recipes in the `plugins/Slimefun/recipes` folder
+are read and added to the `RecipeService`, removing all recipes with the
+same filename. This is why recipes should ideally be *defined* in JSON,
+to prevent unnecessary work.
```
err interesting...

any reason to not just make plugins register their recipes? avoids the whole 1 tick thing.
Also ideally we just don't ever depend on a filename, that's not great.

`RecipeService.registerRecipes(Path.of(getPluginFolder(), "recipes", "recipes.json"));`
(this is reading from addon folder which i also think is fine)

SchnTgaiSpock replied:
I wanted all recipes to be in a single place for the convenience of server owners; like with `Items.yml`

With how custom recipes now work, they will have to be loaded on the first tick too since they take precedence over all other addons' recipes

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+same filename. This is why recipes should ideally be *defined* in JSON,
+to prevent unnecessary work.
+
+When loading JSON recipes, we also need to be able to tell the difference between
```
why do we need to do this? what's the use case?

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+
+While the server is running, recipes can be modified in code, saved to disk, or
+re-loaded from disk. New recipes can also be added, however not to any existing
+file (unless forced, which is not recommended)
```
why not?

SchnTgaiSpock replied:
i also forgot why 😅 the new system should handle this fine i think so i removed it

WalshyDev commented on docs/adr/0002-recipe-rewrite.md:
```diff
@@ -0,0 +1,254 @@
+
+### Stage 4
+
+On server shutdown (or `/sf recipe save`), **all** recipes are saved to disk.
```
Does this include ones where addons register them _without_ a file? If so, why?
I like owners being able to change but I'd also like to not _force_ that to be the case.

SchnTgaiSpock replied:
If an addon registers a new recipe in code it will have to provide a filename, otherwise we wouldn't know where to save it and server owners would have no way to override it. Also disallowing owners from overriding a recipe doesn't do much since they can fork the addon and change it there

WalshyDev reviewed (commented):
Some questions and comments my side but very excited for this, thank you!

SchnTgaiSpock committed:
Update docs/adr/0002-recipe-rewrite.md

Co-authored-by: Daniel Walsh <walshydev@gmail.com>

SchnTgaiSpock committed:
Update docs/adr/0002-recipe-rewrite.md

Co-authored-by: Daniel Walsh <walshydev@gmail.com>

SchnTgaiSpock committed:
Apply suggestions from code review

Co-authored-by: Daniel Walsh <walshydev@gmail.com>

SchnTgaiSpock committed:
update recipe adr

SchnTgaiSpock committed:
Merge branch 'api/recipe-rewrite-4' of https://github.com/SchnTgaiSpock/Slimefun4 into api/recipe-rewrite-4

SchnTgaiSpock committed:
more recipe adr changes

SchnTgaiSpock commented:
@WalshyDev I updated the ADR with most of your review

For custom server recipes, that design was pretty confusing, so I replaced it with one where developers' recipes are in `plugins/Slimefun/recipes` as usual, and custom server recipes are in `plugins/Slimefun/recipes/custom`, which takes precedent

Regarding your comments about recipe filenames -- can you elaborate on why we can't rely on filenames for recipes? It makes overriding them with custom recipes easier. Afaik resource packs also use paths/filenames to override certain resources

Lmk if I missed any parts

SchnTgaiSpock committed:
recipe adr changes

SchnTgaiSpock committed:
recipe adr remove energy and time

SchnTgaiSpock committed:
recipe adr
